pagefaultgames/pokerogue
# [Bug] Move Restriction Battler Tag bugs

## What are the changes the user will see?
The user should be able to use Torment / Gorilla Tactics / Imprison after a refresh.

## Why am I making these changes?
My moves... my responsibility
Addresses #4605 and some other bugs. 

## What are the changes from a developer perspective?
The game is limited in its loading of Battler Tags, leading to bugs. This PR is separate from a complete refactor of Imprison though. 

### Gorilla Tactics
Gorilla Tactics' unique class variable moveId was not included when the tag was loaded in loadTag(). This meant that the moveId would be set to undefined and upon a session reload, none of the Pokemon's moves would understandably match, forcing the Pokemon to Struggle. 

### Torment and Imprison
Torment and Imprison used private class variables to track a Pokemon's active status; however, the Pokemon object is not so easily translatable during a reload. In addition, the parent isMoveRestricted() function did not have access to BattleScene, meaning that it was impossible to regain access to the tracked Pokemon.
This was fixed by adding a new variable representing the affected move user to isMoveRestricted() and its callers. An helper function was added to make retrieving the Pokemon easier. 
#### Imprison
Imprison did not interrupt PRE_MOVE actions correctly. This was fixed by adding a conditional statement that checked for LAPSE_TYPE. 
Imprison can only target the opposing side now. 

### Screenshots/Videos
Gorilla Tactics

https://github.com/user-attachments/assets/5ea3a95d-6d0f-4aca-9d6a-705f325c2472

## How to test the changes?
Gorilla Tactics
1. Use move until next battle.
2. Save and quit.
3. Reload battle. Next battle should be wild. 
4. Should be restricted to the same move.

Torment
1. Enemy Pokemon uses Torment
2. Defeat enemy Pokemon
3. Next battle should be wild. Save and quit + reload battle. 
4. Should be able to use moves. Can't track move history after a reload but it won't freeze.

Imprison
1. Use Imprison
2. Defeat enemy Pokemon
3. Next battle should be wild.  Save and quit + reload battle.
4. Enemy Pokemon should still be affected by Imprison as long as source is present.

## Checklist
